src/toast/ToastState.ts

```diff
@@ -0,0 +1,143 @@
+import React from "react";
+import { v4 as uuidv4 } from "uuid";
+import { Placements } from "./ToastProvider";
+
+type JSXFunction = (props: any) => JSX.Element;
+type StringOrElement = string | JSXFunction;
+
+export interface IToast {
+  id: string;
+  type?: string;
+  content: StringOrElement;
+  timeout?: number;
+  placement?: Placements;
+  autoDismiss?: boolean;
+  isVisible?: boolean;
+}
+
+export type ToastList = Record<string, IToast>;
+
+type GetToastToRenderType = (
+  defaultPlacement: Placements,
+  callback: (position: Placements, toastList: IToast[]) => void,
+) => Array<any>;
+
+export type ToastStateReturn = ReturnType<typeof useToastState>;
+
+const useToastState = ({
+  animationTimeout = 200,
+}: {
+  animationTimeout?: number;
+}) => {
+  const [toasts, setToasts] = React.useState<ToastList>({});
+
+  // toggle can be used to just hide/show the toast instead of removing it.
+  // used for animations, since we don't want to unmount the component directly
+  const toggle = React.useCallback(
+    ({ id, isVisible }: { id: string; isVisible: boolean }) => {
+      setToasts(queue => ({
+        ...queue,
+        [id]: {
+          ...queue[id],
+          isVisible,
+        },
+      }));
+    },
+    [],
+  );
+
+  const show = React.useCallback(
+    ({
+      type = "",
+      content,
+      timeout,
+      autoDismiss,
+      placement,
+    }: Omit<IToast, "id" | "isVisible">) => {
+      const uid = uuidv4();
+      /*
+        wait until the next frame so we can animate
+        wierd bug while using CSSTrasition
+        works fine without RAF & double render when using react spring.
+        maybe because of this:- https://youtu.be/mmq-KVeO-uU?t=842
+       */
+      requestAnimationFrame(() => {
+        setToasts(toasts => ({
+          ...toasts,
+          [uid]: {
+            type,
+            id: uid,
+            content,
+            timeout,
+            placement,
+            autoDismiss,
+            isVisible: false,
+          },
+        }));
+
+        // causes rerender in order to trigger
+        // the animation after mount in CSSTrasition
+        toggle({ id: uid, isVisible: true });
+      });
+    },
+    [toggle],
+  );
+
+  const remove = React.useCallback((id: string) => {
+    // need to use callback based setState otherwise
+    // the remove function would take the queue as dependency
+    // and cause render when changed which would effectively
+    // cause the animations to behave strangly.
+    setToasts(queue => {
+      const newQueue = { ...queue };
+      delete newQueue[id];
+
+      return newQueue;
+    });
+  }, []);
+
+  const hide = React.useCallback(
+    (id: string) => {
+      toggle({ id, isVisible: false });
+
+      window.setTimeout(() => {
+        remove(id);
+      }, animationTimeout);
```

Comment:
Yes it is needed actually.

```js
     window.setTimeout(() => {
        remove(id);
      }, animationTimeout);
```

removes the toast from list after the animation finishes, & ToastController's useTimeout is the actual timeout duration,

There are two timeouts we need one the plain `timeout` which just removes the item after 5sec.

Another for animation since react would instantly unmount the component when we click the close button or it reaches it timeout, so we have to wait till the animation completes and remove the toast after that

Reply:
Got it I understood now.

Can we do one thing here, Get the Timeout ID in a ref and remove them on unmount ?

Reply:
Not needed